### ISSUE TITLE
Platform under rom not 100% 

### ISSUE THREAD

**Issue by Njbmods** [cizia64/CrossMix-OS#200]
When searching for a rom or going through my favourites list, the emulator doesn't show under the game name which is kind of annoying. The stock os had it and it worked great.
Really missing that feature, would be awesome to have a toggle for it.
Cheers!

**Comment by Nevrdid** [comment]
The favorites file is buged. Here is a command/patch to fix it. Tho the issue looks to not comming from crossmix as out of the sort script we never touch the favorite file. #189

**Comment by Njbmods** [comment]
Thanks heaps! Appreciate the time 😊

**Comment by Nevrdid** [comment]
We noticed yesterday on discord that some themes use empty sublabels which is causing the favourites to not display system names. Gonna write a global fix for favorites within the day ^^'

**Comment by Njbmods** [comment]
Absolute legend, i figured it out and my OCD is better now lol. Super work mate, fantastic job on all this! 👏

**Comment by PedroVenancio** [comment]
Hi @Nevrdid 

Here with CrossMix 1.3.0.3 I'm having problems with the emulator identification in Favourites.

First all the games were associated with CPS1. After running System Tools - Tools - Favorites - Fix No Emus I'm getting "wrongbeef" and "Emulators"

![Image](https://github.com/user-attachments/assets/f829694c-1d03-4edc-a862-69f8c94a4661)

am I doing something wrong?

Thanks!

**Comment by Nevrdid** [comment]
Yeah, fav logic is broken more than expected, probably some more fix to add. Can you tho try to overwrite the favorite2.json file by latest timestamp one in same folder ( `Roms/`) and try to add a `sync` between line 16 and 17 in fix script to see if it helps ? Not currently able to test myself ^^
If you can share your fav file here I may see better what the issue.

**Comment by Nevrdid** [comment]
Yeah, fav logic is broken more than expected, probably some more fix to add. Can you tho try to overwrite the favorite2.json file by latest timestamp one in same folder ( `Roms/`) and try to add a `sync` between line 16 and 17 in fix script to see if it helps ? Not currently able to test myself ^^
If you can share your fav file here I may see better what the issue.

**Comment by PedroVenancio** [comment]
Hi @Nevrdid 

Thanks for the quick replay!

It seems that something is duplicating the `sublabel` in [favourite2.json](https://github.com/user-attachments/files/18488067/favourite2.json) 

You can see that the first favourite is fine:

`{"label":"Cadillacs and Dinosaurs (World 930201)","launch":"/mnt/SDCARD/Emus/CPS1/default.sh","rompath":"/mnt/SDCARD/Emus/CPS1/../../Roms/CPS1/dino.zip","imgpath":"/mnt/SDCARD/Emus/CPS1/../../Imgs/CPS1/dino.png","sublabel":"CPS1","type":23}`

but the second has the sublabel repeated:

`{"sublabel":"CPS1","label":"Street Fighter II: The World Warrior (World 910522)","launch":"/mnt/SDCARD/Emus/CPS1/default.sh","rompath":"/mnt/SDCARD/Emus/CPS1/../../Roms/CPS1/sf2.zip","imgpath":"/mnt/SDCARD/Emus/CPS1/../../Imgs/CPS1/sf2.png","sublabel":"CPS1","type":23}`

and the third 

{"sublabel":"CPS1","sublabel":"CPS1","label":"Street Fighter II': Champion Edition (World 920513)","launch":"/mnt/SDCARD/Emus/CPS1/default.sh","rompath":"/mnt/SDCARD/Emus/CPS1/../../Roms/CPS1/sf2ce.zip","imgpath":"/mnt/SDCARD/Emus/CPS1/../../Imgs/CPS1/sf2ce.png","sublabel":"CPS1","type":23}

and so on, until the last

`{"sublabel":"CPS1","sublabel":"CPS1","sublabel":"CPS1","sublabel":"AMIGA","sublabel":"ATARI LYNX","sublabel":"CPS1","sublabel":"CPS1","sublabel":"CPS1","sublabel":"CPS1","sublabel":"CPS1","sublabel":"CPS1","sublabel":"CPS2","sublabel":"CPS2","sublabel":"CPS2","sublabel":"CPS2","sublabel":"CPS2","sublabel":"CPS2","sublabel":"CPS2","sublabel":"CPS3","sublabel":"CPS3","sublabel":"CPS3","sublabel":"DOS","sublabel":"DOS","sublabel":"DOS","sublabel":"N. GBA","sublabel":"N. GBA","sublabel":"N. GBA","sublabel":"N. GBA","sublabel":"N. GBA","sublabel":"N. GBA","sublabel":"N. GBA","sublabel":"N. GBA","sublabel":"N. N64","sublabel":"N. N64","sublabel":"N. NDS","sublabel":"N. NDS","sublabel":"N. SNES","sublabel":"N. SNES","sublabel":"N. SNES","sublabel":"N. SNES","sublabel":"N. SNES","sublabel":"N. SNES","sublabel":"N. SNES","sublabel":"NEOGEO","sublabel":"NEOGEO","sublabel":"OpenBOR","sublabel":"PORTS","sublabel":"PORTS","sublabel":"PORTS","sublabel":"PORTS","sublabel":"PORTS","sublabel":"PORTS","sublabel":"PORTS","sublabel":"PORTS","sublabel":"PORTS","sublabel":"SEGA 32X","sublabel":"SEGA DC","sublabel":"ZXSPECTRUM","sublabel":"ARC. FBNEO","sublabel":"N. GBA","sublabel":"N. GBA","sublabel":"N. GBA","sublabel":"N. GBA","sublabel":"N. GBA","sublabel":"PS","sublabel":"SEGA DC","sublabel":"DOS","sublabel":"SEGA SATURN","sublabel":"SEGA DC","sublabel":"PSP","sublabel":"PSP","sublabel":"ARC. FBNEO","sublabel":"ARC. FBNEO","sublabel":"N. GBA","label":"16. Lotus Esprit Turbo Challenge","launch":"/mnt/SDCARD/Emus/ZXS/default.sh","rompath":"/mnt/SDCARD/Emus/ZXS/../../Roms/ZXS/16. Lotus Esprit Turbo Challenge.7z","imgpath":"/mnt/SDCARD/Emus/ZXS/../../Imgs/ZXS/16. Lotus Esprit Turbo Challenge.png","sublabel":"ZXSPECTRUM","type":23}`

This explains why I'm getting `CPS1` in all favourites:

![Image](https://github.com/user-attachments/assets/3cfd1ac9-6fbf-4aef-8310-f4e6aedbace2)

**Comment by Nevrdid** [comment]
Yup, that exactly what the sed line is supposed to fix ^^' may you have a another favourite2.json without timestamp but an additional letter like favourite2.jsinE. also did you try adding sync after sed ?

**Comment by PedroVenancio** [comment]
Yes, I added `sync` to

```
cd $ROMS_DIR
cp $FAV_FILE "${FAV_FILE}_$timestamp"

sed -i 's/^{.*,"label/{"label/' "$FAV_FILE"
sync
awk '{
  match ($0, /sublabel":"([^"]*)"/, sublabel)
  match ($0, /rompath":"([^"]*)"/, rompath)
  if (sublabel[1] ~ /^[[:space:]]*$/) {
    match(rompath[1], /.*\/Roms\/([^\/]+)\/.*/, emudir)
    gsub(/sublabel":"[^"]*"/, "sublabel\":\"" emudir[1] "\"")
  }
  print
}' $FAV_FILE >favourite2_fixed.json
mv favourite2_fixed.json $FAV_FILE
```

but no luck.

In fact, after run Favorites - Fix No Emus (with or without `sync`) I'm getting "wrongbeef" and "Emulators" in the interface, because the
[favourite2.json](https://github.com/user-attachments/files/18488892/favourite2.json)

have an empy `sublabel` in all favourites:

```
{"label":"Cadillacs and Dinosaurs (World 930201)","launch":"/mnt/SDCARD/Emus/CPS1/default.sh","rompath":"/mnt/SDCARD/Emus/CPS1/../../Roms/CPS1/dino.zip","imgpath":"/mnt/SDCARD/Emus/CPS1/../../Imgs/CPS1/dino.png","sublabel":"","type":23}
{"label":"Street Fighter II: The World Warrior (World 910522)","launch":"/mnt/SDCARD/Emus/CPS1/default.sh","rompath":"/mnt/SDCARD/Emus/CPS1/../../Roms/CPS1/sf2.zip","imgpath":"/mnt/SDCARD/Emus/CPS1/../../Imgs/CPS1/sf2.png","sublabel":"","type":23}
{"label":"Street Fighter II': Champion Edition (World 920513)","launch":"/mnt/SDCARD/Emus/CPS1/default.sh","rompath":"/mnt/SDCARD/Emus/CPS1/../../Roms/CPS1/sf2ce.zip","imgpath":"/mnt/SDCARD/Emus/CPS1/../../Imgs/CPS1/sf2ce.png","sublabel":"","type":23}
```

But the repetitions disapear.

When I change something in favourites (add or remove some game), the repetitions came again.

Also I don't have any favourite2.json without timestamp or with another extension.

Thanks @Nevrdid !

**Comment by Nevrdid** [comment]
So sed work as expected but awk fail ^^' honestly don't see any explanation about why the awk logic could fail on the example you provided. I'll need to try it myself to understand this evening.

**Comment by PedroVenancio** [comment]
@Nevrdid 

Can you please check with

```
awk '{
    # Remove any duplicate sublabel entries first
    gsub(/"sublabel":"[^"]*",/, "")
    
    # Extract the rompath
    if (match($0, /"rompath":"([^"]+)"/, rompath)) {
        # Extract the system name from between /Roms/ and the next /
        if (match(rompath[1], /\/Roms\/([^\/]+)\//, emudir)) {
            # Add a single sublabel after the opening brace
            sub(/^{/, "{\"sublabel\":\"" emudir[1] "\",")
        }
    }
    print
}' $FAV_FILE >favourite2_fixed.json
```

It seems to work here with that change in awk.
